### PR TITLE
Document `LinearSVM`

### PR DESCRIPTION
**Pull request by rcurtin**
This is the last of the simple mlpack classifiers.  Rendered documentation can be seen here:

https://www.ratml.org/misc/mlpack-markdown-doc/linear_svm.html

A number of changes had to happen to the class to bring it in line with other classes:

 * Constructors had to be overhauled to match `LogisticRegression` and `SoftmaxRegression`, which also allow specifying ensmallen optimizers.
 * Versions of `Train()` that allow hyperparameters to be specified were added.
 * Some bugs found when using SGD-style separable optimizers were fixed.
 * The template parameter of the class now represents the storage format of the model, instead of always using `arma::mat`.
 * Tests added to exercise all new support.

### CODE REVIEW

File: src/mlpack/methods/linear_svm/linear_svm_impl.hpp [hunk 1]
```diff
@@ -182,17 +323,30 @@ void LinearSVM<MatType>::Classify(
...
+    typename LinearSVM<ModelMatType>::DenseColType& probabilities) const
+{
+  arma::Row<size_t> labelRow(1);
+  Classify(point, labelRow, probabilities);
+  label = labelRow[0];
+}
+
+template<typename ModelMatType>
+template<typename MatType>
+double LinearSVM<ModelMatType>::ComputeAccuracy(
```
Comment (shrit) [inline src/mlpack/methods/linear_svm/linear_svm_impl.hpp:349]: Can we make the return value `ElemType` please?

File: src/mlpack/methods/linear_svm/linear_svm_impl.hpp [hunk 2]
```diff
@@ -54,107 +83,218 @@ LinearSVM<MatType>::LinearSVM(
...
+  const bool needNewParameters = (parameters.is_empty() ||
+      (fitIntercept && (parameters.n_rows != data.n_rows + 1)) ||
+      (!fitIntercept && (parameters.n_rows != data.n_rows)) ||
+      (parameters.n_cols != numClasses));
+  if (needNewParameters)
     parameters = svm.InitialPoint();
 
   // Train the model.
-  const double out = optimizer.Optimize(svm, parameters);
+  const double out = optimizer.Optimize(svm, parameters, callbacks...);
```
Comment (shrit) [inline src/mlpack/methods/linear_svm/linear_svm_impl.hpp:261]: We need to change `double` to `ElemType` in here as well, since we are returning `ElemType`

**Review by shrit — APPROVED** [review]
Just a couple of comments that we might need to fix before merging, otherwise everything is looking very good :+1:

File: src/mlpack/methods/linear_svm/linear_svm_impl.hpp (hunk 1 above)
Comment (rcurtin) [reply src/mlpack/methods/linear_svm/linear_svm_impl.hpp:349]: (See also [this related comment/response](https://github.com/mlpack/mlpack/pull/3572#discussion_r1431934932).)  I don't mind returning the accuracy as a `double`; it won't have any real computational implications, and makes it so that when the user changes `ElemType`, there are fewer methods whose signatures change.

File: src/mlpack/methods/linear_svm/linear_svm_impl.hpp (hunk 2 above)
Comment (rcurtin) [reply src/mlpack/methods/linear_svm/linear_svm_impl.hpp]: Thanks, nice catch!  Fixed in 621ade26. :+1:

**Review by zoq — APPROVED** [review]

**Review by shrit — APPROVED** [review]
